### PR TITLE
MM-14384 ephemeral post function demo

### PR DESCRIPTION
**Pull request by scottleedavis**
Fixes https://github.com/mattermost/mattermost-server/issues/10392

* ephemeral post function demo.  adds `/demo_plugin ephemeral`
* Update to 5.10 mattermost-server dependency

### CODE REVIEW

File: server/command_hooks.go [hunk 1]
```diff
@@ -17,10 +17,10 @@ func (p *Plugin) registerCommand(teamId string) error {
 		TeamId:           teamId,
 		Trigger:          CommandTrigger,
 		AutoComplete:     true,
-		AutoCompleteHint: "(true|false)",
-		AutoCompleteDesc: "Enables or disables the demo plugin hooks.",
+		AutoCompleteHint: "(true|false|ephemeral)",
+		AutoCompleteDesc: "Enables or disables the demo plugin hooks, or demonstrates an ephemeral post capabilities.",
 		DisplayName:      "Demo Plugin Command",
-		Description:      "A command used to enable or disable the demo plugin hooks.",
+		Description:      "A command used to enable or disable the demo plugin hooks, or demonstrate ephemeral post capabilities.",
```
Comment (hanzei) [inline server/command_hooks.go]: I would like to add this as a new command. 3/5
Maybe just `/ephemeral`? 1/5

File: server/http_hooks.go [hunk 2]
```diff
@@ -45,3 +51,66 @@ func (p *Plugin) handleStatus(w http.ResponseWriter, r *http.Request) {
...
+		Message:   fmt.Sprintf("updated ephemeral action %d", count),
+		Props: model.StringInterface{
+			"attachments": []*model.SlackAttachment{
+				{
+					Actions: []*model.PostAction{
+						{
+							Id: model.NewId(),
+							Integration: &model.PostActionIntegration{
+								Context: model.StringInterface{
+									"count": count,
```
Comment (hanzei) [inline server/http_hooks.go]: :100:

File: server/http_hooks.go [hunk 3]
```diff
@@ -45,3 +51,66 @@ func (p *Plugin) handleStatus(w http.ResponseWriter, r *http.Request) {
...
 }
+
+func (p *Plugin) handleEphemeralUpdate(w http.ResponseWriter, r *http.Request) {
+	request := model.PostActionIntegrationRequestFromJson(r.Body)
+	URL := fmt.Sprintf("%s", *p.API.GetConfig().ServiceSettings.SiteURL)
+
+	count := int(request.Context["count"].(float64)) + 1
+
+	post := &model.Post{
+		Id:        request.PostId,
```
Comment (hanzei) [inline server/http_hooks.go]: I don't think this is needed. I would rather not be explicit here, as most integrations arn't either.

File: server/http_hooks.go [hunk 4]
```diff
@@ -45,3 +51,66 @@ func (p *Plugin) handleStatus(w http.ResponseWriter, r *http.Request) {
...
+		ChannelId: request.ChannelId,
+		UserId:    request.UserId,
```
Comment (hanzei) [inline server/http_hooks.go]: Same here and with `ChannelId`.

File: server/http_hooks.go [hunk 5]
```diff
@@ -45,3 +51,66 @@ func (p *Plugin) handleStatus(w http.ResponseWriter, r *http.Request) {
...
+						{
+							Id: model.NewId(),
+							Integration: &model.PostActionIntegration{
+								Context: model.StringInterface{
+									"count": count,
+								},
+								URL: fmt.Sprintf("%s/plugins/%s/ephemeral/update", URL, manifest.Id),
+							},
+							Type: model.POST_ACTION_TYPE_BUTTON,
+							Name: "Update",
```
Comment (hanzei) [inline server/http_hooks.go]: Maybe show `count` in this? 0/5

File: server/command_hooks.go [hunk 6]
```diff
@@ -82,6 +82,45 @@ func (p *Plugin) ExecuteCommand(c *plugin.Context, args *model.CommandArgs) (*mo
...
+		post := &model.Post{
+			ChannelId: args.ChannelId,
+			UserId:    args.UserId,
+			Message:   "test ephemeral actions",
+			Props: model.StringInterface{
+				"attachments": []*model.SlackAttachment{
+					{
+						Actions: []*model.PostAction{
+							{
+								Id: model.NewId(),
```
Comment (hanzei) [inline server/command_hooks.go]: Is this needed? I don't think so.

File: server/http_hooks.go [hunk 7]
```diff
@@ -45,3 +51,66 @@ func (p *Plugin) handleStatus(w http.ResponseWriter, r *http.Request) {
...
+			},
+		},
+	}
+	p.API.UpdateEphemeralPost(request.UserId, post)
+
+	resp := &model.PostActionIntegrationResponse{}
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	_, _ = w.Write(resp.ToJson())
+
```
Comment (hanzei) [inline server/http_hooks.go]: Nit: Remove this line.

File: server/http_hooks.go [hunk 8]
```diff
@@ -45,3 +51,66 @@ func (p *Plugin) handleStatus(w http.ResponseWriter, r *http.Request) {
...
+}
+
+func (p *Plugin) handleEphemeralDelete(w http.ResponseWriter, r *http.Request) {
+	request := model.PostActionIntegrationRequestFromJson(r.Body)
+
+	post := &model.Post{
+		Id: request.PostId,
+	}
+	p.API.DeleteEphemeralPost(request.UserId, post)
```
Comment (hanzei) [inline server/http_hooks.go]: This is 100% correct. But it looks awful. We have to fix this API method some day...

File: go.sum [hunk 9]
```diff
@@ -44,6 +46,8 @@ github.com/magiconair/properties v1.8.0 h1:LLgXmsheXeRoUOBOjtwPQCWIYqM/LU1ayDtDe
 github.com/magiconair/properties v1.8.0/go.mod h1:PppfXfuXeibc/6YijjN8zIbojt8czPbwD3XqdrwzmxQ=
 github.com/mattermost/mattermost-server v5.9.0+incompatible h1:bUlZKAhjp7cGQRepmmsv225vbXyZaXOuVstqgpB+cE4=
 github.com/mattermost/mattermost-server v5.9.0+incompatible/go.mod h1:5L6MjAec+XXQwMIt791Ganu45GKsSiM+I0tLR9wUj8Y=
```
[the remark this answers is not in the document]
Comment (scottleedavis) [reply go.sum]: Done.  :)

File: server/command_hooks.go (hunk 1 above)
Comment (scottleedavis) [reply server/command_hooks.go]: Created a new command, and felt `demo_ephemeral` was a better candidate.  Open to changes.

File: server/http_hooks.go (hunk 2 above)
Comment (scottleedavis) [reply server/http_hooks.go]: :)

File: server/http_hooks.go (hunk 3 above)
Comment (scottleedavis) [reply server/http_hooks.go]: Id is needed so UpdateEphemeralPost knows which post to update.  (unless you meant count, which is used to show how many times the post was updated) .  Could you clarify what you meant please?

File: server/http_hooks.go (hunk 4 above)
Comment (scottleedavis) [reply server/http_hooks.go]: Without `ChannelId` , UpdateEphemeralPost doesn't appear to work for my tests.

File: server/http_hooks.go (hunk 5 above)
Comment (scottleedavis) [reply server/http_hooks.go]: I really like this idea.  Changed to put count in the update button.

File: server/command_hooks.go (hunk 6 above)
Comment (scottleedavis) [reply server/command_hooks.go]: Not needed.  Removed Id from each `PostAction`

File: server/http_hooks.go (hunk 7 above)
Comment (scottleedavis) [reply server/http_hooks.go]: the header?  removed on both that and in handleEphemeralDelete.

File: server/http_hooks.go (hunk 8 above)
Comment (scottleedavis) [reply server/http_hooks.go]: 100% agreed.  I noticed you brought it up in ~extensions earlier today, so I look forward to the new version.

File: server/http_hooks.go (hunk 3 above)
Comment (scottleedavis) [reply server/http_hooks.go]: Ah, I think you mean't count.  I updated it to this:
`count := request.Context["count"].(float64) + 1`

File: server/http_hooks.go (hunk 4 above)
Comment (levb) [reply server/http_hooks.go]: That's true, I ran into the same thing. Worth a ticket, or by design?

Comment (scottleedavis) [reply server/http_hooks.go]: As ChannelId is usually readily available, I am 0/5 if it stays in.  That being said, it seems PostId may be enough by itself?

File: server/command_hooks.go (hunk 1 above)
Comment (hanzei) [reply server/command_hooks.go]: In #32 `/crash` is used and it quite like this. I'm 3/5 on `/ephemeral` now. It's a short and clear name for what the plugin does. Why do you want to go with `demo_ephemeral`?

File: server/http_hooks.go (hunk 4 above)
Comment (hanzei) [reply server/http_hooks.go]: After thinking a bit more about this, I would leave it as it is. The only require informations are `Id` and `ChannelId`, correct? That seams fine to me.

File: server/http_hooks.go [hunk 10]
```diff
@@ -45,3 +51,62 @@ func (p *Plugin) handleStatus(w http.ResponseWriter, r *http.Request) {
...
+									"count": count,
+								},
+								URL: fmt.Sprintf("%s/plugins/%s/ephemeral/update", URL, manifest.Id),
+							},
+							Type: model.POST_ACTION_TYPE_BUTTON,
+							Name: fmt.Sprintf("Update %d", int(count)),
+						},
+						{
+							Integration: &model.PostActionIntegration{
+								Context: model.StringInterface{},
```
Comment (hanzei) [inline server/http_hooks.go]: This isn't needed. Mind deleting it?

File: server/command_hooks.go [hunk 11]
```diff
@@ -44,44 +56,77 @@ func (p *Plugin) emitStatusChange() {
...
+										"count": 0,
+									},
+									URL: fmt.Sprintf("%s/plugins/%s/ephemeral/update", URL, manifest.Id),
+								},
+								Type: model.POST_ACTION_TYPE_BUTTON,
+								Name: "Update",
+							},
+							{
+								Integration: &model.PostActionIntegration{
+									Context: model.StringInterface{},
```
Comment (hanzei) [inline server/command_hooks.go]: This isn't needed. Mind deleting it?

File: server/command_hooks.go [hunk 12]
```diff
@@ -44,44 +56,77 @@ func (p *Plugin) emitStatusChange() {
...
+			ChannelId: args.ChannelId,
+			UserId:    args.UserId,
+			Message:   "test ephemeral actions",
+			Props: model.StringInterface{
+				"attachments": []*model.SlackAttachment{
+					{
+						Actions: []*model.PostAction{
+							{
+								Integration: &model.PostActionIntegration{
+									Context: model.StringInterface{
```
Comment (hanzei) [inline server/command_hooks.go]: I would like (2/5) go get rid of `model.StringInterface` in this code base. What do you think about using `map[string]interface{}` instead?

File: server/http_hooks.go [hunk 13]
```diff
@@ -45,3 +51,62 @@ func (p *Plugin) handleStatus(w http.ResponseWriter, r *http.Request) {
...
+					},
+				},
+			},
+		},
+	}
+	p.API.UpdateEphemeralPost(request.UserId, post)
+
+	resp := &model.PostActionIntegrationResponse{}
+	w.WriteHeader(http.StatusOK)
+	_, _ = w.Write(resp.ToJson())
```
Comment (hanzei) [inline server/http_hooks.go]: Is this need? Isn't an empty response enough?

File: server/http_hooks.go [hunk 14]
```diff
@@ -45,3 +51,62 @@ func (p *Plugin) handleStatus(w http.ResponseWriter, r *http.Request) {
...
+
+}
+
+func (p *Plugin) handleEphemeralDelete(w http.ResponseWriter, r *http.Request) {
+	request := model.PostActionIntegrationRequestFromJson(r.Body)
```
Comment (hanzei) [inline server/http_hooks.go]: See above

File: server/http_hooks.go (hunk 8 above)
Comment (hanzei) [reply server/http_hooks.go]: Well, this got fixed faster then I thought. What do you think about adapting the new API and requiring MM `v5.12.0` for this plugin.

File: server/command_hooks.go (hunk 1 above)
Comment (scottleedavis) [reply server/command_hooks.go]: I had originally wanted to preserve the 'demo_*' namespace, but if we are going to add `/crash` it makes sense to use `/ephemeral` too.

File: server/http_hooks.go (hunk 8 above)
Comment (scottleedavis) [reply server/http_hooks.go]: Sounds good.  I tried
```	github.com/mattermost/mattermost-server v5.12.0+incompatible``` in go.mod, but there is an error  `unknown revision v5.12.0`.  Leaving as is, and changing `const minimumServerVersion = "5.12.0"`

Comment (scottleedavis) [reply server/http_hooks.go]: Seems that using minimumServerVersion at 5.12 isn't available yet on master & release-5.12 branches haven't updated version yet.  

Also, when i update go.mod/go.sum to the latest mattermost master, I get these seemingly unrelated errors
```
# github.com/mattermost/mattermost-plugin-demo/server [github.com/mattermost/mattermost-plugin-demo/server.test]
server/configuration.go:189:28: p.Helpers undefined (type *Plugin has no field or method Helpers)
server/configuration_test.go:71:28: undefined: plugintest.Helpers
server/configuration_test.go:86:26: undefined: plugintest.Helpers
server/configuration_test.go:87:17: undefined: plugintest.Helpers
server/configuration_test.go:106:26: undefined: plugintest.Helpers
server/configuration_test.go:107:17: undefined: plugintest.Helpers
server/configuration_test.go:124:26: undefined: plugintest.Helpers
server/configuration_test.go:125:17: undefined: plugintest.Helpers
server/configuration_test.go:145:5: p.SetHelpers undefined (type Plugin has no field or method SetHelpers)
```

What would you like me to do @hanzei ?  Update and fix the changes, or leave this to be fixed in a future PR?  (or other) . happy to do whichever.  :)

(paused changing to the new DeleteEphemeralPost until this is resolved.)

File: server/http_hooks.go [hunk 15]
```diff
@@ -45,3 +51,62 @@ func (p *Plugin) handleStatus(w http.ResponseWriter, r *http.Request) {
...
+func (p *Plugin) handleEphemeralUpdate(w http.ResponseWriter, r *http.Request) {
+	request := model.PostActionIntegrationRequestFromJson(r.Body)
+	URL := fmt.Sprintf("%s", *p.API.GetConfig().ServiceSettings.SiteURL)
+
+	count := request.Context["count"].(float64) + 1
+
+	post := &model.Post{
+		Id:        request.PostId,
+		ChannelId: request.ChannelId,
+		UserId:    request.UserId,
```
[the remark this answers is not in the document]
Comment (scottleedavis) [reply server/http_hooks.go]: Appears not.   Removing.

File: server/http_hooks.go [hunk 16]
```diff
@@ -45,3 +51,62 @@ func (p *Plugin) handleStatus(w http.ResponseWriter, r *http.Request) {
 func (p *Plugin) handleHello(w http.ResponseWriter, r *http.Request) {
 	w.Write([]byte("Hello World!"))
 }
+
+func (p *Plugin) handleEphemeralUpdate(w http.ResponseWriter, r *http.Request) {
+	request := model.PostActionIntegrationRequestFromJson(r.Body)
+	URL := fmt.Sprintf("%s", *p.API.GetConfig().ServiceSettings.SiteURL)
```
[the remark this answers is not in the document]
Comment (scottleedavis) [reply server/http_hooks.go]: Habit I suppose.  using your recommendation.

File: server/http_hooks.go [hunk 17]
```diff
@@ -45,3 +51,62 @@ func (p *Plugin) handleStatus(w http.ResponseWriter, r *http.Request) {
 func (p *Plugin) handleHello(w http.ResponseWriter, r *http.Request) {
 	w.Write([]byte("Hello World!"))
 }
+
+func (p *Plugin) handleEphemeralUpdate(w http.ResponseWriter, r *http.Request) {
+	request := model.PostActionIntegrationRequestFromJson(r.Body)
```
[the remark this answers is not in the document]
Comment (scottleedavis) [reply server/http_hooks.go]: Done

File: server/command_hooks.go (hunk 11 above)
Comment (scottleedavis) [reply server/command_hooks.go]: Done

File: server/command_hooks.go (hunk 12 above)
Comment (scottleedavis) [reply server/command_hooks.go]: Using model.StringInterface looks cleaner to me personally.  I can make the switch to `map[string]interface{}` if you'd prefer though.

File: server/http_hooks.go (hunk 14 above)
Comment (scottleedavis) [reply server/http_hooks.go]: Done

File: server/http_hooks.go (hunk 13 above)
Comment (scottleedavis) [reply server/http_hooks.go]: Without this response errors appear in the log `Action integration error                      caller="mlog/log.go:172" err_details="err=EOF" err_where=DoPostAction http_code=400 ip_addr="::1" method=POST path=/api/v4/posts/5qz78hib7idpzdo9fun5i4jypo/actions/74377u6mfbbk9kr38orwu1319w request_id=njzege7kcjrqicw7mfbnkzende user_id=mywge9a3ibyd9nosbkaaubo4ph`

File: server/command_hooks.go (hunk 12 above)
Comment (hanzei) [reply server/command_hooks.go]: I'm only 2/5. Let's go with `model.StringInterface`, if you prefer it.

File: server/http_hooks.go (hunk 13 above)
Comment (hanzei) [reply server/http_hooks.go]: Didn't know that. Thanks!

File: server/configuration.go [hunk 18]
```diff
@@ -220,7 +220,7 @@ func (p *Plugin) ensureDemoUser(configuration *configuration) (string, error) {
 	if user == nil {
 		user, err = p.API.CreateUser(&model.User{
 			Username:  configuration.Username,
-			Password:  "password",
+			Password:  "P@ssw0rd!!",
```
Comment (hanzei) [inline server/configuration.go]: Why is this change needed? :question:

File: server/http_hooks.go (hunk 8 above)
Comment (hanzei) [reply server/http_hooks.go]: You can update the dependency with `go get github.com/mattermost/mattermost-server@2d3e41783398dc16689e5c1732f9d21727383e2c`. Would you please try this?


We can bump `minimumServerVersion` once DB bump is done. This should be done in the next couple of days.

Comment (scottleedavis) [reply server/http_hooks.go]: That works perfectly @hanzei 👍

File: server/configuration.go (hunk 18 above)
Comment (scottleedavis) [reply server/configuration.go]: Whoops.  I was testing and on a branch there was a 10 character + special characters limit.  Changing this back

File: server/README.md [hunk 19]
```diff
@@ -59,6 +59,9 @@ This demo implementation logs a message to the demo channel whenever a user leav
 This demo implementation responds to a `/demo_plugin` command, allowing the user to enable
 or disable the demo plugin's hooks functionality (but leave the command and webapp enabled).
 
+A third option 'ephemeral' demonstrates ephemeral interactive usage of SendEphemeralPost,
```
Comment (hanzei) [inline server/README.md]: This has to be updated. Sorry, forgot it in the last review.

Comment (scottleedavis) [reply server/README.md]: Done.  👍

**Comment by hanzei** [comment]
Thanks again Scott for this contribution :+1:
